f3/participant.go

```diff
@@ -49,6 +49,12 @@ func (p *Participant) ReceiveCanonicalChain(chain ECChain, power PowerTable, bea
 	}
 }
 
+func (p *Participant) ReceiveECChain(chain ECChain) {
+	if p.granite != nil && chain.HasPrefix(p.granite.input) {
+		p.granite.heaviestChain = chain
```

Comment:
Please don't mutate the granite instance externally. Instead, add a method to receive a new chain and enforce the prefix logic there.

The predicate isn't right. It must check that the current heaviest acceptable chain is a prefix, not just the input. Otherwise this can switch chains mid-execution.